include/pthread.h

```diff
@@ -0,0 +1,376 @@
+ *
+ * Note that pthread attribute structs are currently noops in Zephyr.
+ */
+static inline int pthread_condattr_destroy(pthread_condattr_t *att)
+{
+	return 0;
+}
+
+/**
+ * @brief Declare a pthread mutex
+ *
+ * Declaration API for a pthread mutex.  This is not a POSIX API, it's
+ * provided to better conform with Zephyr's allocation strategies for
+ * kernel objects.
+ *
+ * @param name Symbol name of the mutex
+ */
+#define PTHREAD_MUTEX_DEFINE(name)		\
+	K_SEM_DEFINE(name##_psem, 1, 1);	\
+	struct pthread_mutex name = {		\
+		.sem = &name##_psem,		\
+	}
+
+
+/**
+ * @brief POSIX threading compatibility API
+ *
+ * See IEEE 1003.1
+ */
+static inline int pthread_mutex_init(pthread_mutex_t *m,
+				     const pthread_mutexattr_t *att)
+{
+	ARG_UNUSED(att);
+
+	k_sem_init(m->sem, 1, 1);
+
+	return 0;
+}
+
+/**
+ * @brief POSIX threading compatibility API
+ *
+ * See IEEE 1003.1
+ */
+static inline int pthread_mutex_destroy(pthread_mutex_t *m)
+{
+	ARG_UNUSED(m);
+
+	return 0;
+}
+
+/**
+ * @brief POSIX threading compatibility API
+ *
+ * See IEEE 1003.1
+ */
+static inline int pthread_mutex_lock(pthread_mutex_t *m)
+{
+	return k_sem_take(m->sem, K_FOREVER);
+}
+
+/**
+ * @brief POSIX threading compatibility API
+ *
+ * See IEEE 1003.1
+ */
+static inline int pthread_mutex_timedlock(pthread_mutex_t *m,
+					  const struct timespec *to)
+{
+	int ret = k_sem_take(m->sem, _ts_to_ms(to));
+
+	return ret == -EAGAIN ? -ETIMEDOUT : ret;
+}
+
+/**
+ * @brief POSIX threading compatibility API
+ *
+ * See IEEE 1003.1
+ */
+int pthread_mutex_trylock(pthread_mutex_t *m);
+
+/**
+ * @brief POSIX threading compatibility API
+ *
+ * See IEEE 1003.1
+ */
+static inline int pthread_mutex_unlock(pthread_mutex_t *m)
+{
+	k_sem_give(m->sem);
+	return 0;
+}
+
+/**
+ * @brief POSIX threading compatibility API
+ *
+ * See IEEE 1003.1
+ *
+ * Note that pthread attribute structs are currently noops in Zephyr.
+ */
+static inline int pthread_mutexattr_init(pthread_mutexattr_t *m)
+{
+	ARG_UNUSED(m);
+
+	return 0;
+}
+
+/**
+ * @brief POSIX threading compatibility API
+ *
+ * See IEEE 1003.1
+ *
+ * Note that pthread attribute structs are currently noops in Zephyr.
+ */
+static inline int pthread_mutexattr_destroy(pthread_mutexattr_t *m)
+{
+	ARG_UNUSED(m);
+
+	return 0;
+}
+
+/* FIXME: these are going to be tricky to implement.  Zephyr has (for
+ * good reason) deprecated its own "initializer" macros in favor of a
+ * static "declaration" macros instead.  Using such a macro inside a
+ * gcc compound expression to declare and object then reference it
+ * would work, but gcc limits such expressions to function context
+ * (because they may need to generate code that runs at assignment
+ * time) and much real-world use of these initializers is for static
+ * variables.  The best trick I can think of would be to declare it in
+ * a special section and then initialize that section at runtime
+ * startup, which sort of defeats the purpose of having these be
+ * static...
```

Comment:
You can have functions marked with `__attribute__((constructor))`, which is essentially declaring it in a special section. We already support calling these functions when `CONFIG_CPLUSPLUS` is enabled (see `_main()` in `init.c`).